### PR TITLE
[Wallet] Refactoring by using CInputCoin instead of std::pair

### PR DESCRIPTION
**Pull request by NicolasDorier**
The code was using `std::pair<const CWalletTx*, unsigned int>` until now. Not only it is ugly to read, and easy to mess up, but it needlessly pass `CWalletTx` all around the code when it is not needed.

This code will simplify another PR extending `FundRawTransaction`. (https://github.com/bitcoin/bitcoin/pull/10068)

This is divided in 3 commits easy to review.

1. Just rename `std::pair<const CWalletTx*, unsigned int>` to `CInputCoin` by using a type alias
2. Transform `CInputCoin` into a class.
3. Rewrite some code which can be simplified thanks to `CInputCoin`

### CODE REVIEW

**Comment by laanwj** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/10165/commits/2f5eb873dda9d1561c233a2313322a53afea5161

File: src/wallet/wallet.h [hunk 1]
```diff
@@ -475,7 +475,38 @@ class CWalletTx : public CMerkleTx
 };
 
 
-using CInputCoin = std::pair<const CWalletTx*, unsigned int>;
+class CInputCoin {
+public:
+    CInputCoin()
```
Comment (ryanofsky) [inline src/wallet/wallet.h]: Would suggest deleting default constructor and IsNull method. Their presence seem to make this class more complex and precarious than is necessary, and could lead to bugs and unintended behavior.

The only place where a nullable `CInputCoin` seems to be used is in one `coinLowestLarger` local variable, and you could more easily declare this variable as `boost::optional<CInputCoin> coinLowestLarger;` (or [std::optional](http://en.cppreference.com/w/cpp/utility/optional) in the future) instead of building in a layer of optionality into the type.

**Review by ryanofsky — COMMENTED** [review]
utACK 2f5eb873dda9d1561c233a2313322a53afea5161, nice cleanup

File: src/wallet/wallet.h (hunk 1 above)
Comment (NicolasDorier) [inline src/wallet/wallet.h]: I don't think using `optional` would make the code more clear. I never saw it used anywhere in bitcoin.
What about using `CInputCoin*` for coinLowestLarger instead?

Comment (ryanofsky) [inline src/wallet/wallet.h]: > I don't think using optional would make the code more clear.

I think optional's useful. It saves work of having to build optionality into types you define, and more importantly it allows the same type to be optional in some contexts and required in others, which I think is exactly what you want here. std::optional is part of c++17 so eventually it will start becoming more commonly used.

Anyway, this was just a suggestion, so feel free to ignore it if you have a different preference.

> What about using CInputCoin* for coinLowestLarger instead?

Using a pointer was actually going to be my first suggestion, but I didn't see a way to make that work because coinLowestLarger is assigned from a local variable. But I think this would be an improvement if it could work.

**Comment by NicolasDorier** [comment]
I needed to rebase, I ended up using `boost::optional` and getting ride of `IsNull()`.

**Comment by btcdrak** [comment]
utACK 7e0b365

**Comment by laanwj** [comment]
> I needed to rebase, I ended up using boost::optional 

Was about to ask "why not std::optional" but apparently that's C++17. Well, easy to replace when we adapt that - the concept is great.

File: src/wallet/wallet.h [hunk 2]
```diff
@@ -475,7 +475,7 @@ class CWalletTx : public CMerkleTx
 };
 
 
-
+using CInputCoin = std::pair<const CWalletTx*, unsigned int>;
```
Comment (JeremyRubin) [inline src/wallet/wallet.h]: Prefer to use a typedef (I actually prefer using personally, but some devs are unfamiliar with it and you don't need it as CInputCoin is not templated)

Also, prefer a non-Hungarian name if you're introducing a new class.
See https://github.com/bitcoin/bitcoin/pull/9472#discussion_r95775684 -- Hungarians for new code discouraged.

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -2106,22 +2105,21 @@ bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, const int nConfMin
...
         }
-        else if (n < nTargetValue + MIN_CHANGE)
+        else if (coin.txout.nValue < nTargetValue + MIN_CHANGE)
         {
             vValue.push_back(coin);
-            nTotalLower += n;
+            nTotalLower += coin.txout.nValue;
         }
-        else if (n < coinLowestLarger.first)
+        else if (coinLowestLarger.IsNull() || coin.txout.nValue < coinLowestLarger.txout.nValue)
```
Comment (JeremyRubin) [inline src/wallet/wallet.cpp]: I'm a bit confused here -- isn't coinLowestLarger always null at this point?

File: src/wallet/wallet.h [hunk 4]
```diff
@@ -475,7 +475,38 @@ class CWalletTx : public CMerkleTx
...
+        {
+            outpoint = COutPoint(walletTx->GetHash(), i);
+            txout = walletTx->tx->vout[i];
+        }
+    }
+    bool IsNull() const
+    {
+        return outpoint.IsNull() && txout.IsNull();
+    }
+    COutPoint outpoint;
```
Comment (JeremyRubin) [inline src/wallet/wallet.h]: prefer not to have fields intermixed with functions.

File: src/wallet/wallet.h [hunk 5]
```diff
@@ -477,21 +477,21 @@ class CWalletTx : public CMerkleTx
...
 public:
-    CInputCoin()
-    {
-    }
     CInputCoin(const CWalletTx* walletTx, unsigned int i)
     {
-        if (walletTx != nullptr && i < walletTx->tx->vout.size())
+        if (walletTx == nullptr)
+            throw std::invalid_argument("walletTx should not be null");
+        if (i < walletTx->tx->vout.size())
```
Comment (JeremyRubin) [inline src/wallet/wallet.h]: Can you invert the logic here `if (!P) throw` is a bit easier to read.

Might also be good to make it a `const CWalletTx&` so as to make sure that de-referencability is checked everywhere that this is called.

**Review by JeremyRubin — COMMENTED** [review]
Couple nits/questions (be sure to show outdated) but overall Concept Ack.

File: src/wallet/wallet.h [hunk 6]
```diff
@@ -475,7 +475,38 @@ class CWalletTx : public CMerkleTx
...
+            txout = walletTx->tx->vout[i];
+        }
+        else
+        {
+            throw std::out_of_range("The output index is out of range");
+        }
+    }
+
+    COutPoint outpoint;
+    CTxOut txout;
```
Comment (bulldozer00) [inline src/wallet/wallet.h]: Why are these data members public and not private?

File: src/wallet/test/wallet_tests.cpp [hunk 7]
```diff
@@ -29,7 +29,7 @@ extern UniValue importmulti(const JSONRPCRequest& request);
 
 std::vector<std::unique_ptr<CWalletTx>> wtxn;
 
-typedef std::set<std::pair<const CWalletTx*,unsigned int> > CoinSet;
+typedef std::set<CInputCoin> CoinSet;
```
Comment (bulldozer00) [inline src/wallet/test/wallet_tests.cpp]: Think about replacing old style "typedef  X Y " with the more intuitive C++11 "using Y = X" style

Comment (NicolasDorier) [inline src/wallet/test/wallet_tests.cpp]: better to do in a separate PR. I want to keep it focused on `CInputCoin`.

File: src/wallet/wallet.h (hunk 6 above)
Comment (NicolasDorier) [inline src/wallet/wallet.h]: because I use them outside the class.

File: src/wallet/wallet.h (hunk 5 above)
Comment (NicolasDorier) [inline src/wallet/wallet.h]: Ah indeed, C# habits... Just replaced by `if(!walletTx)`.

I would like to do `const CWalletTx&` in a separate PR, as it need more code change, and I need this PR for another one depending on it.

Comment (JeremyRubin) [inline src/wallet/wallet.h]: Ah, I was unclear -- I actually don't mind a `== nullptr` (although current state is cleaner), I was referring to the `throw out_of_range`. It's easier if a constructor is: `if (! preconditionA) throw A; if (!preconditionB) throw B; ....;` rather than `if (! preconditionA) throw A; if (preconditionB)...; else throw B;`

Comment (NicolasDorier) [inline src/wallet/wallet.h]: @JeremyRubin indeed, much clearer. I updated the PR.

**Comment by NicolasDorier** [comment]
Travis fail unrelated to this PR, @laanwj can you kick the travis box one time to fix the issue ?

File: src/wallet/test/wallet_tests.cpp (hunk 7 above)
Comment (laanwj) [inline src/wallet/test/wallet_tests.cpp]: We don't use that style anywhere, introducing it here is not a good idea. I'd prefer to save that for new code.

File: src/wallet/wallet.h [hunk 8]
```diff
@@ -475,7 +475,34 @@ class CWalletTx : public CMerkleTx
 };
 
 
-
+class CInputCoin {
```
Comment (laanwj) [inline src/wallet/wallet.h]: It is not visible on github but local diff shows me that line endings are weird here:
```diff
-using CInputCoin = std::pair<const CWalletTx*, unsigned int>;
+class CInputCoin {^M
+public:^M
```
^M is 13, so I suppose these are DOS line endings. Looks like it is only in `src/wallet/wallet.h`, and only the newly  introduced code.
suggest running:
```bash
$ dos2unix src/wallet/wallet.h
dos2unix: converting file src/wallet/wallet.h to Unix format ...
```

Comment (NicolasDorier) [inline src/wallet/wallet.h]: oops. Yes, developping on visual studio... Fixed.
